DOCS-8072 -4.3-tuning-batch-processing-kt (#1421)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-batch-processing.adoc

> = Batch Processing

Reviewer: ```suggestion
= How Batch Processing Improves Processing Performance
```
Maybe many of the titles in these sets should be "How <xyz> Improves Processing Performance"

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Mule runtime engine (Mule) possesses the ability to process messages in batches. Within an application, you can initiate a batch job which is a block of code that splits large messages into individual records, performs actions upon each record, then reports on the results and potentially pushes the processed output to other systems or queues.

Reviewer: Simplify wording; concision:
```suggestion
Mule runtime engine (Mule) can process messages in batches. When your application initiates a batch job, that job splits large messages into individual records, performs actions upon each record, and then reports the results and potentially pushes the processed output to other systems or queues.
```

Reviewer: But...where do "batches" come in to this? Typically, "batching" is the action of combining many small things into one larger thing, for simultaneous processing. This text specifies breaking up a larger thing into smaller things and processing them individually, which sounds like the opposite of batching.

Are we saying that the individual records are processed and then the RESULTS are batched?

> Consider having 1 million records processed by a batch application with three steps. Several I/O operations occur as Mule processes each record while they move through the job’s phases. The disk characteristics, along with the workload size, play a key role in the performance of the batch job. Mainly, because during the input phase, an in-disk queue is created with the list of records to be processed, which is constantly read and written by the flow resulting in heavy I/O utilization.

Reviewer: It is not clear to me how this paragraph relates to the previous one.

Are the 1 million records the result of splitting a large message?

This paragraph refers to "a **batch application**" rather than an **application** in which a **batch job** is initiated; are these the same thing? If so, standardize to use one wording.

It says there are "three steps," but what are those steps? The previous para refers to splitting, actions, and reporting, which are three "steps," I guess, but this one refers to "the job's phases." What are the steps, and what are the phases?

What are "the disk characteristics" that impact batch performance?

How is "the flow" related to batch processing?

If batch processing improves performance, why do we end by mentioning "heavy I/O utilization," which I assume results in poor performance?

Reviewer: Here is a copyedit of existing text...but consider reworking to address the questions I raise above:
```suggestion
Consider having 1 million records processed by a batch application with three steps. Several I/O operations occur as Mule processes each record as it moves through the job’s phases. The disk characteristics, along with the workload size, play a key role in the performance of the batch job, primarily because during the input phase, an in-disk queue is created of the list of records to be processed, which is constantly read and written by the flow, resulting in heavy I/O utilization.
```

> Additionally, batch processing requires having enough available memory to process the threads in parallel, which means moving the records from persistent storage into RAM in a fixed size block. The larger your records and their quantity, the more available memory you need for batch processing.

Reviewer: copyedited only:
```suggestion
Additionally, batch processing requires having enough memory available to process the threads in parallel, which means moving the records from persistent storage into RAM in a fixed-size block. The larger your records and their quantity, the more available memory you need for batch processing.
```
ditto re. inconsistent terminology: "threads" vs. "records"?
Is it clear in the previous text that the records move from persistent storage to RAM, and what is the fixed size of the block? (Note that earlier text uses "block" to refer to the larger code broken into individual records).

> == Batch Block Size

Reviewer: Because the previous text refers to "records in a fixed-size block, you probably can continue the text here without the subtitle.
I am only just now getting the sense that "batch" = "block of records." This needs to be explicit from the beginning. 
Consider defining how "batch," "message," "block," "records," "job", and "application" all fit together from the start of the topic. If there are "steps" or "phases," define them as well. Maybe something like...:

"You can improve application processing performance by using _batch processing_, which takes large message blocks of code and divides them into individual records that are then processed individually and simultaneously rather than in a long linear manner. The results of these processed records are then batching into a single result and reported or forwarded to other systems or queues.

Successful processing of multiple individual records simultaneously requires significant RAM, because records are moved to RAM from persistent storage to speed processing time. To balance the needs of performance and working memory, the default size for a batch job is 100 records, based on analysis across a set of representative batch use cases with various record sizes. You can change this value by using the following syntax if, after running comparative tests with different values, you find that the optimum batch size for your use case is different: 

[source,xml,linenums]
----
<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

By default, the batch block size is set to 100. It is a good balancing point between performance and working memory requirements based on analysis across a set of representative batch use cases with various record sizes. However, the optimal value for each application depends on its use case. +
You can configure this property for the batch job. If you believe that in your particular case, custom size is best suited, make sure to run comparative tests with different values to find the optimum size in your use case. The following example shows the batch block size:

[source,xml,linenums]
----
<batch:job jobName="test-batch" blockSize="${batch.block.size}">

> ----

Reviewer: See above.

> To take full advantage of this feature, you must understand how the block sizes affect your batch job. Running comparative tests with different values and testing performance helps you find an optimum block size before deploying this change into production.

Reviewer: omit; repetitive
